### ISSUE TITLE
[BUG] Volume cloning retry logic causes unnecessary updates to engine CR

### ISSUE THREAD

**Issue by ejweber** [longhorn/longhorn#8952]
## Describe the bug

After https://github.com/longhorn/longhorn-manager/pull/2895, engine CRs for attached volumes are updated every five seconds, even if nothing is happening with the engines.

## To Reproduce

I discovered this issue while executing a `kubectl -n longhorn-system get engine -w` during a normal development flow. I was surprised to find all engines in my cluster updating every five seconds or so. I initially thought there were no changes except for the `resourceVersion`, but closer inspection from @PhanLe1010 yielded the diff at https://www.diffchecker.com/fAcS8sXw/.

## Expected behavior

If nothing is happening with an engine, its CR should not be updated.

## Environment

 - Longhorn version: `master-head` nearing the `v1.7.0` release (`4d90cc71fb5fbb33008bb9de870f2e8cb018af58`)

**Comment by PhanLe1010** [comment]
## Test Plan

Case 1:
Repeat this test plan https://github.com/longhorn/longhorn/issues/2907#issuecomment-2204026385

Case 2:
1. Create a test volume
1. Attach it
1. Watch the engine CR `k get engine -n longhorn-system -w`
2. Verify that engine CR are not updated repeatedly

**Comment by ejweber** [comment]
Created https://github.com/longhorn/longhorn/issues/8953 to be implemented in `v1.8.0` to help avoid these kinds of accidental issues in response to an offline discussion with @PhanLe1010, @james-munson, and @shuo-wu.

**Comment by roger-ryao** [comment]
**Verified on master-head 20240709**

- longhorn `master-head` https://github.com/longhorn/longhorn/commit/09b1a8ebc79a1c083090bbfd96c64667d77964e0
- longhorn-manager `master-head` https://github.com/longhorn/longhorn-manager/commit/754a93e3ee86bbfcb0325c96d2666b1422556daf


**The test steps**
Case 1:
[the case 1 of the test plan](https://github.com/longhorn/longhorn/issues/2907#issuecomment-2204026385):

1. Create storage class with `replica count = 1`:
```bash
kubectl create -f - <<EOF
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: test-sc
provisioner: driver.longhorn.io
allowVolumeExpansion: true
parameters:
  numberOfReplicas: "1"
EOF
```
2. Create source pvc:
```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-pvc
  namespace: default
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: test-sc
  resources:
    requests:
      storage: 5Gi
EOF
```
3. Create a pod using the volume:
```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  containers:
    - name: ubuntu
      image: ubuntu
      command: ["/bin/sleep", "3650d"]
      volumeMounts:
        - name: pod-data
          mountPath: /data
  volumes:
    - name: pod-data
      persistentVolumeClaim:
        claimName: test-pvc
EOF
``` 
4. Write 4GB data to the volume and calculate the checksum.
```bash
kubectl exec -it test-pod -- /bin/bash -c "dd if=/dev/urandom of=/data/4G bs=1M count=4096 oflag=direct status=progress && md5sum /data/4G"
```
5. Clone the source pvc:
```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: cloned-pvc
spec:
  storageClassName: test-sc
  dataSource:
    name: test-pvc
    kind: PersistentVolumeClaim
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
EOF
```
6. While volume cloning is in progress, delete the `instance manager` on the same node of the source pvc replica.
7. Verify that Longhorn should fail the clone. Then retry. Then the clone should eventually succeeded.
[longhorn-manager-4rttg_longhorn-manager.log](https://github.com/user-attachments/files/16141282/longhorn-manager-4rttg_longhorn-manager.log)

8. After the volume cloning is complete, check the cloned volume data.

```bash
kubectl create -f - <<EOF
apiVersion: v1
kind: Pod
metadata:
  name: test-pod-cloned
spec:
  containers:
    - name: ubuntu
      image: ubuntu
      command: ["/bin/sleep", "3650d"]
      volumeMounts:
        - name: pod-data-cloned
          mountPath: /data
  volumes:
    - name: pod-data-cloned
      persistentVolumeClaim:
        claimName: cloned-pvc
EOF
``` 

Case 2:
1. Create a test volume
1. Attach it
1. Watch the engine CR `k get engine -n longhorn-system -w`
1. Verify that engine CR is not updated repeatedly 

**Result passed**
